lib/syslog/pool/pool.go

```diff
@@ -0,0 +1,92 @@
+package pool
+
+import (
+	"io"
+)
+
+// A LimitedConnPool is a connection pool, with the property that
+// the total number of live connections is limited to the size
+// parameter passed to NewLimited. In order to make this guarantee,
+// we assume that all new connections are introduced by calls to the
+// Get method, and returned to the pool by the Put method (whether
+// or not the connection is dead)
+type LimitedConnPool struct {
+	conns  chan io.WriteCloser // Available connections in the pool
+	live   chan struct{}       // Keep a count of living connections (in our hands, or the client)
+	signal chan struct{}       // Used to wake up the connection producer
+	err    chan error          // Send dial errors back to the client
+}
+
+// NewLimited returns a new LimitedConnPool with the given size limit and dial function.
+func NewLimited(size int, dial func() (io.WriteCloser, error)) (*LimitedConnPool, error) {
+	// Tentative first try - if this doesn't work, we assume it never will
+	// and fail to initialize.
+	w, err := dial()
+	if err != nil {
+		return nil, err
+	}
+
+	p := LimitedConnPool{
+		conns:  make(chan io.WriteCloser, size),
+		live:   make(chan struct{}, size),
+		signal: make(chan struct{}),
+
+		// try to make this large enough to avoid dropping
+		// errors if clients only check errors occasionally
+		err: make(chan error, size),
+	}
+
+	p.conns <- w
+	p.live <- struct{}{}
+
+	// keep p.conns populated
+	go func() {
+		for range p.signal {
+			for len(p.live) < size {
```

Comment:
I think relying on `len(p.live)` could end up causing too many connections to be opened. Let say `cap(p.live)` is 3, you've got all 3 connections acquired, one is lost and then sends a value on `p.signal`. At this point the code will cause 3 new connections to be opened (assuming the 2 remaining haven't been pushed back yet).

There are few ways to resolve this:
1. Maintain a separate counter of active connections to control how many must be created
2. Open as many connections as values we get on the `p.signal` channel, so for each connection lost we re-open exactly one.
2. Use a non-blocking send on the `p.live` channel when giving the connections back, so it never blocks because too many connections have been created (but we would be violating the limit of connections set when creating the pool).

Reply:
I misread the code never mind :+1:

Reply:
(cleared this up with @achille-roussel in person, p.live *is* the active connection counter)